# Setup

In [302]:
import pandas as pd
pd.options.display.float_format = "{:,.2f}".format

begInv = pd.read_csv("BegInvFINAL12-31-16Sample.csv")
endInv = pd.read_csv("EndInvFINAL12-31-16Sample.csv")
purchases = pd.read_csv("PurchasesFINAL12-31-16Sample.csv")
invoices = pd.read_csv("InvoicePurchases12-31-16Sample.csv")
sales = pd.read_csv("SalesFINAL12-31-16Sample.csv")

In [303]:
# begInv = pd.read_csv("BegInvFINAL12312016.csv")
# endInv = pd.read_csv("EndInvFINAL12312016.csv")
# purchases = pd.read_csv("PurchasesFINAL12312016.csv")
# invoices = pd.read_csv("InvoicePurchases12312016.csv")
# sales = pd.read_csv("SalesFINAL12312016.csv")

# Deliverable 1

## A

In [304]:
begInv = begInv.rename(columns = {'onHand':'01_01_2016_Qty'}) #Change column name for later readability when merging

begInv['01_01_2016_InvCost'] = begInv['01_01_2016_Qty'] * begInv['Price']

In [305]:
begInvStore = begInv.groupby("Store").agg({"01_01_2016_Qty": "sum", "01_01_2016_InvCost": "sum"}).sort_values("01_01_2016_InvCost", ascending=False).reset_index()

begInvStore.head(10)

Store  01_01_2016_Qty  01_01_2016_InvCost
0     34          153852        3,291,170.24
1     73          162551        3,142,497.36
2     67          158996        3,079,578.63
3     66          149314        2,973,033.90
4     76          140208        2,952,418.44
5     69          144255        2,946,726.65
6     38          114368        2,232,698.77
7     55          119641        2,001,263.66
8     50           94720        1,649,808.22
9     79           95330        1,503,149.48

## B

In [306]:
# begInvBrand = begInv.groupby(["InventoryId", "Description"]).agg({"01_01_2016_Qty": "sum", "01_01_2016_InvCost": "sum"}).sort_values("01_01_2016_InvCost", ascending=False).reset_index()

begInvBrand = begInv.groupby(["Brand"]).agg({"01_01_2016_Qty": "sum", "01_01_2016_InvCost": "sum"}).sort_values("01_01_2016_InvCost", ascending=False).reset_index()

begInvBrand.head(10)

Brand  01_01_2016_Qty  01_01_2016_InvCost
0   3545           14499          463,823.01
1   1233           12016          432,455.84
2   8068           15341          383,371.59
3   4261           15499          340,823.01
4   3858           13649          327,439.51
5   2753            4625          286,703.75
6   8082            9287          278,517.13
7   8680            7066          275,503.34
8   2589            6766          270,572.34
9   3876           14829          266,773.71

# Deliverable 2

## A

In [307]:
endInv = endInv.rename(columns = {'onHand':'12_31_2016_Qty'}) #Change column name for later readability when merging

endInv['12_31_2016_InvCost'] = endInv['12_31_2016_Qty'] * endInv['Price']

In [308]:
endInvStore = endInv.groupby("Store").agg({"12_31_2016_Qty": "sum", "12_31_2016_InvCost": "sum"}).sort_values("12_31_2016_InvCost", ascending=False).reset_index()

endInvStore.head(10)

Store  12_31_2016_Qty  12_31_2016_InvCost
0     50          260717        4,887,260.68
1     73          164589        3,254,662.81
2     67          163765        3,076,114.82
3     34          145829        3,074,616.75
4     76          143866        2,975,945.18
5     69          150848        2,968,678.82
6     66          144579        2,860,504.99
7     74          166015        2,803,645.13
8     38          129397        2,463,906.85
9     55          125584        2,234,836.35

## B

In [309]:
endInvBrand = endInv.groupby(["Brand"]).agg({"12_31_2016_Qty": "sum", "12_31_2016_InvCost": "sum"}).sort_values("12_31_2016_InvCost", ascending=False).reset_index()

endInvBrand.head(10)

Brand  12_31_2016_Qty  12_31_2016_InvCost
0   1233           15047          526,494.53
1   3545           16770          502,932.30
2   2753            7849          470,861.51
3   8068           15608          366,631.92
4   3405           12268          355,649.32
5   4261           16769          351,981.31
6   2757           11603          336,370.97
7   2589            7922          300,956.78
8   1376           13180          276,648.20
9   2585           10487          272,557.13

# Deliverable 3

In [310]:
totalInv = pd.merge(begInv, endInv, how="outer", on=["InventoryId"])

# totalInv = totalInv[["Brand", "01_01_2016_Qty", "01_01_2016_InvCost", "12_31_2016_Qty", "12_31_2016_InvCost"]].rename(columns={"01_01_2016_Qty": "Beginning Quantity", "01_01_2016_InvCost": "Beginning Total", "12_31_2016_Qty": "Ending Quantity", "12_31_2016_InvCost": "Ending Total"}).sort_values("Brand", ascending=True).reset_index()

totalInv = totalInv[["InventoryId", "01_01_2016_Qty", "01_01_2016_InvCost", "12_31_2016_Qty", "12_31_2016_InvCost"]]

totalInv = totalInv.rename(columns={"01_01_2016_Qty": "Beginning Quantity", "01_01_2016_InvCost": "Beginning Total", "12_31_2016_Qty": "Ending Quantity", "12_31_2016_InvCost": "Ending Total"})

totalInv.fillna(0, inplace=True)

totalInv[["Beginning Quantity", "Ending Quantity"]] = totalInv[["Beginning Quantity", "Ending Quantity"]].astype(int)

totalInv.head()

InventoryId  Beginning Quantity  Beginning Total  Ending Quantity   
0  1_HARDERSFIELD_58                   8           103.92               11  \
1  1_HARDERSFIELD_60                   7            76.93                0   
2  1_HARDERSFIELD_62                   6           221.94                7   
3  1_HARDERSFIELD_63                   3           116.97                7   
4  1_HARDERSFIELD_72                   6           209.94                4   

   Ending Total  
0        142.89  
1          0.00  
2        258.93  
3        272.93  
4        139.96

In [311]:
len(totalInv)

256042

# Deliverable 4

## A

In [312]:
#Filters only 2016 data
purchases2016 = purchases[purchases["InvoiceDate"].str.startswith("2016")]

In [313]:
topVend = purchases2016.groupby(["VendorNumber", "VendorName"]).agg({"Dollars": "sum"}).sort_values("Dollars", ascending=False).reset_index()

topVend.head(10)

VendorNumber                   VendorName       Dollars
0          3960  DIAGEO NORTH AMERICA INC    20,777,975.89
1          4425        MARTIGNETTI COMPANIES 11,019,308.94
2         12546  JIM BEAM BRANDS COMPANY      9,711,692.24
3         17035  PERNOD RICARD USA            9,709,368.10
4           480  BACARDI USA INC              7,283,824.24
5          1392  CONSTELLATION BRANDS INC     6,416,659.81
6          1128  BROWN-FORMAN CORP            5,539,694.61
7          9165  ULTRA BEVERAGE COMPANY LLP   5,341,404.57
8          3252  E & J GALLO WINERY           5,002,534.49
9          9552  M S WALKER INC               4,480,443.78

### For those using Invoice Data or full purchase data

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th></th>
      <th>Dollars</th>
    </tr>
    <tr>
      <th>VendorNumber</th>
      <th>VendorName</th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3960</th>
      <th>DIAGEO NORTH AMERICA INC</th>
      <td>49,733,333.66</td>
    </tr>
    <tr>
      <th>4425</th>
      <th>MARTIGNETTI COMPANIES</th>
      <td>26,810,799.70</td>
    </tr>
    <tr>
      <th>12546</th>
      <th>JIM BEAM BRANDS COMPANY</th>
      <td>23,539,441.79</td>
    </tr>
    <tr>
      <th>17035</th>
      <th>PERNOD RICARD USA</th>
      <td>23,460,507.21</td>
    </tr>
    <tr>
      <th>480</th>
      <th>BACARDI USA INC</th>
      <td>16,985,053.64</td>
    </tr>
    <tr>
      <th>1392</th>
      <th>CONSTELLATION BRANDS INC</th>
      <td>15,187,741.32</td>
    </tr>
    <tr>
      <th>1128</th>
      <th>BROWN-FORMAN CORP</th>
      <td>13,139,921.12</td>
    </tr>
    <tr>
      <th>9165</th>
      <th>ULTRA BEVERAGE COMPANY LLP</th>
      <td>12,433,070.97</td>
    </tr>
    <tr>
      <th>3252</th>
      <th>E &amp; J GALLO WINERY</th>
      <td>11,858,245.15</td>
    </tr>
    <tr>
      <th>9552</th>
      <th>M S WALKER INC</th>
      <td>10,294,481.73</td>
    </tr>
  </tbody>
</table>
</div>

## B

In [314]:
#Recognized on Invoice Date
invoices_2016ID = invoices[invoices["InvoiceDate"].str.startswith("2016")]

#Generates an overall list of said values in groupby and agg()
topVendAll = invoices_2016ID.groupby(["VendorNumber", "VendorName"]).agg({"Dollars": "sum", "Freight": "sum", "Quantity": "sum"}).reset_index()

#Generates a filtered list of just the requested values
topVendCharge = topVendAll[["VendorNumber", "VendorName", "Freight"]].sort_values("Freight", ascending=False).reset_index(drop=True)

topVendCharge.head(10)

VendorNumber                   VendorName    Freight
0          3960  DIAGEO NORTH AMERICA INC    250,899.75
1          4425        MARTIGNETTI COMPANIES 139,767.62
2         12546  JIM BEAM BRANDS COMPANY     120,230.57
3         17035  PERNOD RICARD USA           120,130.51
4           480  BACARDI USA INC              86,153.58
5          1392  CONSTELLATION BRANDS INC     77,791.20
6          1128  BROWN-FORMAN CORP            66,576.22
7          9165  ULTRA BEVERAGE COMPANY LLP   63,888.38
8          3252  E & J GALLO WINERY           59,594.41
9          9552  M S WALKER INC               52,199.20

## C

In [315]:
# #Generates new column to hold the rate
# topVendAll["Freight Per $ Purchased"] = (topVendAll["Freight"] / topVendAll["Dollars"]) 

# #Filters out all the rows not >=250k
# topVend250 = topVendAll[topVendAll["Dollars"] >= 250000].copy(deep=True) 

# topVend250 = topVend250[["VendorNumber", "VendorName", "Freight Per $ Purchased"]].sort_values("Freight Per $ Purchased", ascending=False).reset_index(drop=True)

# # topVend250["Freight Per $ Purchased"] = topVend250["Freight Per $ Purchased"].round(4).map("${:.4f}".format)


In [316]:
#Filters out all the rows not >=250k
topVend250 = topVendAll[topVendAll["Dollars"] >= 250000].copy(deep=True) 

#Generates new column to hold the rate
topVend250["FreightRate$(%)"] = (topVend250["Freight"] / topVend250["Dollars"])*100
topVend250["FreightPerUnit"] = (topVend250["Freight"] / topVend250["Quantity"])

topVend250 = topVend250[["VendorNumber", "VendorName", "Freight", "Dollars", "Quantity", "FreightRate$(%)","FreightPerUnit"]].sort_values("FreightRate$(%)", ascending=False).reset_index(drop=True)

In [317]:
topVend250.head(10)

VendorNumber                   VendorName    Freight       Dollars   
0          9625  WESTERN SPIRITS BEVERAGE CO   1,746.69    323,948.73  \
1          1590  DIAGEO CHATEAU ESTATE WINES   7,083.48  1,331,574.55   
2          9744  FREDERICK WILDMAN & SONS      3,885.16    736,944.55   
3         17031  FLAG HILL WINERY & VINEYARD   1,477.16    281,173.13   
4           516  BANFI PRODUCTS CORP           8,034.30  1,529,676.79   
5           653      STATE WINE & SPIRITS      7,757.25  1,481,052.81   
6          4425        MARTIGNETTI COMPANIES 139,767.62 26,810,799.70   
7          8892  TRINCHERO FAMILY ESTATES     11,225.72  2,175,075.75   
8          8673  STE MICHELLE WINE ESTATES    15,507.50  3,007,382.16   
9          9815  WINE GROUP INC               26,280.87  5,109,629.23   

   Quantity  FreightRate$(%)  FreightPerUnit  
0     52243             0.54            0.03  
1    183317             0.53            0.04  
2     68087             0.53            0.06  
3     19188             0.53            0.08  
4    213967             0.53            0.04  
5    149590             0.52            0.05  
6   2544805             0.52            0.05  
7    340040             0.52            0.03  
8    411345             0.52            0.04  
9    863638             0.51            0.03

In [318]:
# topVend250.tail(5)

In [319]:
topVend250.sort_values("FreightRate$(%)", ascending=True).head()

VendorNumber                   VendorName   Freight      Dollars   
52          6359  OLE SMOKY DISTILLERY LLC     1,876.61   377,754.50  \
51          9819  TREASURY WINE ESTATES       14,470.47 2,903,971.85   
50          3089  SIDNEY FRANK IMPORTING CO    8,258.10 1,656,429.63   
49          3924  HEAVEN HILL DISTILLERIES    13,854.16 2,773,519.92   
48          8352  LUXCO INC                    9,785.35 1,951,124.00   

    Quantity  FreightRate$(%)  FreightPerUnit  
52    108698             0.50            0.02  
51    487654             0.50            0.03  
50    179772             0.50            0.05  
49    346500             0.50            0.04  
48    271443             0.50            0.04

## D

In [320]:
#Filter out from the 2016 dataframe to take requested values
d4Master = invoices_2016ID[(invoices_2016ID["Freight"] > 100) & (invoices_2016ID["Quantity"] <= 1000)].copy(deep=True)

# d4_d["Freight Per $ Purchased"] = (d4_d["Freight"] / d4_d["Dollars"])
d4_d = d4Master.groupby(["VendorNumber", "VendorName"]).agg({"Dollars": "sum", "Quantity": "sum", "Freight": "sum"}).reset_index().sort_values("Freight", ascending=False)

# d4_d = d4_d.groupby(["VendorNumber", "VendorName", ]).agg({"Dollars": "sum", "Freight": "sum", "Quantity": "sum", "Freight Per $ Purchased": "sum"}).reset_index()

# d4_d = d4_d.groupby(["VendorNumber", "VendorName",  "Freight Per $ Purchased"]).agg({"Dollars": "sum", "Freight": "sum", "Quantity": "sum"}).reset_index()
d4_d = d4_d[["VendorNumber", "VendorName", "Freight", "Dollars", "Quantity"]].reset_index(drop=True)
d4_d.head()

VendorNumber                   VendorName  Freight    Dollars  Quantity
0          2561  EDRINGTON AMERICAS          2,944.67 570,627.52     18120
1          8664  STOLI GROUP,(USA) LLC         349.02  14,299.98       975
2          7239  REMY COINTREAU USA INC        348.07   7,515.38       382
3           653      STATE WINE & SPIRITS      303.70   5,544.89       541
4          6785  PALM BAY INTERNATIONAL INC    242.73   5,533.18       715

## E

In [321]:
d4_d["Freight Per $ Purchased (E)"] = (d4_d["Freight"] / d4_d["Dollars"])
d4_d["Freight Per Unit"] = (d4_d["Freight"] / d4_d["Quantity"])

d4_e = d4_d[["VendorNumber", "Freight Per $ Purchased (E)", "Freight Per Unit"]]
d4_e = pd.merge(topVend250, d4_e, how="inner", on=["VendorNumber"]).sort_values("Freight Per $ Purchased (E)", ascending=False)

d4_e = d4_e.reset_index(drop=True).drop(d4_e.columns[0], axis=1) # remove both index columns
d4_e.head()


VendorName   Freight      Dollars  Quantity   
0      STATE WINE & SPIRITS     7,757.25 1,481,052.81    149590  \
1  REMY COINTREAU USA INC      16,369.66 3,230,350.59    136450   
2  PALM BAY INTERNATIONAL INC  13,324.60 2,634,237.03    367789   
3  DISARONNO INTERNATIONAL LLC  5,719.22 1,113,007.23    100060   
4  MHW LTD                      9,282.22 1,826,631.59    209630   

   FreightRate$(%)  FreightPerUnit  Freight Per $ Purchased (E)   
0             0.52            0.05                         0.05  \
1             0.51            0.12                         0.05   
2             0.51            0.04                         0.04   
3             0.51            0.06                         0.04   
4             0.51            0.04                         0.04   

   Freight Per Unit  
0              0.56  
1              0.91  
2              0.34  
3              0.38  
4              0.43

In [322]:
#Average rate from Deliverable E/D
# d4_e["Freight Per $ Purchased (E)"].mean()

In [323]:
#Average rate from Deliverable C
# d4_e["Freight Per $ Purchased"].mean()

Comparing Deliverable C and E, the freight cost per $ purchased was found to be much cheaper in larger purchases (>$250k) versus the higher overall cost in E/D which only looked at freight costs >$100. Therefore, on average, large total purchases offer better freight per dollar purchased as the data has indicated when compared to overall freight charges above $100.

# Deliverable 5


In [324]:
mergeSales = sales.groupby("InventoryId").agg({"SalesQuantity": "sum", "SalesDollars": "sum"}).sort_values("SalesQuantity", ascending=False).reset_index()

In [325]:
# Merge sales inventory on the inventory ID column
mergeSales = pd.merge(mergeSales, totalInv, how="outer", on=["InventoryId"])

In [326]:
# Calculate purchases by subtracting the beginning inventory quantity from the ending inventory quantity, and then adding the sales quantity
mergeSales["Purchases"] = (mergeSales["Ending Quantity"] - mergeSales["Beginning Quantity"]) + mergeSales["SalesQuantity"]
mergeSales.fillna(0, inplace=True)
mergeSales[["SalesQuantity", "Beginning Quantity", "Ending Quantity", "Purchases"]] = mergeSales[["SalesQuantity", "Beginning Quantity", "Ending Quantity","Purchases"]].astype(int)

In [327]:
mergeSales = mergeSales.sort_values("Purchases", ascending=False).reset_index(drop=True)
mergeSales = mergeSales[["InventoryId", "SalesQuantity", "Beginning Quantity", "Ending Quantity", "Purchases"]]

In [328]:
mergeSales.head()

InventoryId  SalesQuantity  Beginning Quantity  Ending Quantity   
0   50_MOUNTMEND_5609             10                   0             3676  \
1  67_EANVERNESS_5111            286                 164             3005   
2   50_MOUNTMEND_5614             10                   0             1778   
3  74_PAENTMARWY_8111           1750                 271              246   
4   50_MOUNTMEND_1233           1009                 537             1246   

   Purchases  
0       3686  
1       3127  
2       1788  
3       1725  
4       1718

This dataframe utilizes the beginning and ending tables to generate a table that outputs the number of purchases made on a inventory item to indicate both popularity and total number of replenishment for that product done throughout the 2016 period.